In [475]:
import pandas as pd
from Bio.Seq import Seq
from Bio import Entrez
import os
from typing import List, Dict

In [476]:
os.chdir('C:\\Users\\jcham\spanins')
os.getcwd()

'C:\\Users\\jcham\\spanins'

## Goal: Return CDS for 'sequence', given 'ncbi_sequence' and 'cds' (ncbi cds).
* Note: The spanin sequences in general are subsets of one another

# Data exploration and prep

### Table dimensions are 122x13

In [477]:
data = pd.read_csv('category_I_sequence_diff.csv')
display(data.head(8), data.shape)

host  host_taxid    phage_acc     protein_acc  protein_gi  \
0       Aeromonas         642  NC_020879.1  YP_007677913.1   472438133   
1       Aeromonas         642  NC_019543.1  YP_007010874.1   423262275   
2  Enterobacteria         547     AP011113             NaN    26042140   
3  Enterobacteria         547  NC_024142.1  YP_009031980.1   640884671   
4   Campylobacter         194  NC_016562.1  YP_004956969.1   371671013   
5     Escherichia         561     EU078592             NaN    40093751   
6     Pseudomonas         286  NC_007805.1  YP_001293406.1   148912827   
7      Salmonella         590   KC139515.1      AGF88018.1   451937721   

              cds    strand function spanin_type property_feature  \
0  126555..126789  NEGATIVE  ISPANIN         sep        TMD:7..29   
1  124571..124826  NEGATIVE  ISPANIN         sep        TMD:7..29   
2             NaN       NaN  ISPANIN         sep        TMD:5..24   
3    12841..13066  POSITIVE  ISPANIN         emb       TMD:10..27   
4    89202..89547  NEGATIVE  ISPANIN         ovl       TMD:15..34   
5             NaN       NaN  ISPANIN         emb        TMD:4..26   
6    37910..38360  POSITIVE  ISPANIN         emb              NaN   
7    17634..17832  NEGATIVE  ISPANIN         ovl        TMD:4..26   

                                            sequence  \
0  MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1  MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
2  MHVSNFTAGLLLLVIAFGGTSIILKNKVERLETSVVEITKTANENA...   
3  MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   
5  MSRVTAIISALVICIIVCLSWAVNHYRDNAITYKAQRDKNARELKL...   
6  MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   
7  MTGLLARIKTGVLAALVFVVALFGVWRAGRTKGKQDQINNQNNDTL...   

                                       ncbi_sequence same  
0  MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...    N  
1  MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...    N  
2                                                NaN  NaN  
3  MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...    N  
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...    N  
5                                                NaN  NaN  
6  MVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQ...    N  
7  MALFGVWRAGRTKGKQDQINNQNNDTLREQANADKNVAEVHNEINK...    N

(122, 13)

### NaN rows
* 10 of the NaN rows do not have any ncbi_sequence data (probably weren't called at all by the submitter, but were called by rohit) **Need to manual pblast these probably?**
* The other 10 do have sequence data but are missing other data
* Rows 6, 8, 9, 15, 20, 28, and 32 are missing only the property_feature data. These rows should be manageable with the planned script
* Rows 45, 79 and 99 have sequence data but no cds or strand. Probably need to be dealt with manually. 
* All others remaining (2, 5, 25, 40, 41, 42, 74, 81, 114, 118) were not called by the ncbi submitter.

In [478]:
data[data.isna().any(axis=1)]

host  host_taxid    phage_acc     protein_acc  \
2              Enterobacteria         547     AP011113             NaN   
5                 Escherichia         561     EU078592             NaN   
6                 Pseudomonas         286  NC_007805.1  YP_001293406.1   
8                  Salmonella         590  NC_010392.1  YP_001700615.1   
9                  Salmonella         590  NC_010393.1  YP_001700673.1   
15                Escherichia         561  NC_024379.1  YP_009044294.1   
20               Burkholderia       32008  NC_009234.1  YP_001111073.1   
25                   Yersinia         629     HE956707             NaN   
28                   Yersinia         629     AM076770      CAJ28448.1   
32           Stenotrophomonas       40323  NC_023588.1  YP_009008371.1   
40   Escherichia coli O157:H7         561     KM092515             NaN   
41             Enterobacteria         547  NC_011045.1             NaN   
42             Enterobacteria         547     AP011113             NaN   
45              Campylobacter         194   FN667789.1      CBJ94317.1   
74                     vibrio         662  NC_005083.2             NaN   
79               Edwardsiella         635  NC_021342.1             NaN   
81               Bdellovibrio         662  NC_015721.1             NaN   
99             Enterobacteria         547    NC_012223             NaN   
114  Escherichia coli O157:H7         561     KM092515             NaN   
118                    vibrio         662  NC_023863.2             NaN   

     protein_gi           cds    strand function spanin_type property_feature  \
2      26042140           NaN       NaN  ISPANIN         sep        TMD:5..24   
5      40093751           NaN       NaN  ISPANIN         emb        TMD:4..26   
6     148912827  37910..38360  POSITIVE  ISPANIN         emb              NaN   
8     169257238  30405..30858  NEGATIVE  ISPANIN         emb              NaN   
9     169257297  17493..17973  POSITIVE  ISPANIN         emb              NaN   
15    658607293  36334..36736  POSITIVE  ISPANIN         emb              NaN   
20    134288770  32853..33276  NEGATIVE  ISPANIN         ovl              NaN   
25    398313030           NaN       NaN  ISPANIN         ovl       TMD:10..27   
28    164414553  18144..18336  POSITIVE  ISPANIN         ovl              NaN   
32    589892004    9447..9939  POSITIVE  ISPANIN         emb              NaN   
40     26631539           NaN       NaN  ISPANIN         ovl       TMD:10..27   
41     93201455           NaN       NaN  OSPANIN         emb      LIPO:29..32   
42     26042139           NaN       NaN  OSPANIN         sep      LIPO:14..17   
45    294338279           NaN       NaN  OSPANIN         sep      LIPO:15..18   
74      2546051           NaN       NaN  OSPANIN         sep      LIPO:14..17   
79    509141582           NaN       NaN  OSPANIN         ovl      LIPO:16..19   
81     33882695           NaN       NaN  OSPANIN         ovl      LIPO:25..28   
99    225220121           NaN       NaN  OSPANIN         ovl      LIPO:14..17   
114    26631540           NaN       NaN  OSPANIN         ovl      LIPO:17..20   
118   657201235           NaN       NaN  USPANIN         USP      LIPO:14..17   

                                              sequence  \
2    MHVSNFTAGLLLLVIAFGGTSIILKNKVERLETSVVEITKTANENA...   
5    MSRVTAIISALVICIIVCLSWAVNHYRDNAITYKAQRDKNARELKL...   
6    MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   
8    MMFNWKTMFVGLLLVSLIVAGRLANHYRNNAITYKYQRDTATHNLK...   
9    MMFNWKTMFVGLLLVSLIVAGRLANHYRNNAITYKEQRDTVTHRLT...   
15   MLEFLKRAAPWLLAAVMFAGGYHTANNKWEAKVNAEYTSNLKASED...   
20   MNLSRLMPWLALFALIALAASCQHGRALRAQLERATDDARRANRDA...   
25   MLTIPNKYKWAVMALLAAVSIGSLTLANHYRDSALTSQKALQEVTD...   
28   MLGKLKIAVMLMIAAVLAWKAGSWNGARVERSVQIAECNNRIEKLA...   
32   MLYRALALAALVLATAGLFSYQQGRISRATTALDKANLDLAKARSE...   
40   MTFKEFCQATFIIVFLVGAGVWGGYSYRDYQVAEAELSNEKLISVA...   
41   MKYRLSIKKTLPRWKGALIGLFLICVATISGCVSESKLPEPPMVSV...   
42   MIKLSAVIL

### Creating data2 DF, which drops irrelevant columns and 13 NaN rows
* ```data2``` = original data minus rows with NaN values for sequence, ncbi_sequence or cds columns
* ```data2``` contains 7 more rows than simply running dropna() on ```data``` (as some rows have NaN values in irrelevant columns)


In [479]:
data2 = data[['phage_acc', 'strand', 'cds', 'sequence', 'ncbi_sequence']].copy()
data2 = data2.dropna(axis=0)
display(data2.head(3), data2.shape)

phage_acc    strand             cds  \
0  NC_020879.1  NEGATIVE  126555..126789   
1  NC_019543.1  NEGATIVE  124571..124826   
3  NC_024142.1  POSITIVE    12841..13066   

                                            sequence  \
0  MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1  MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
3  MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   

                                       ncbi_sequence  
0  MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...  
1  MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...  
3  MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...

(109, 5)

### Creating list of the protein sequences (rohit's spanin calls vs NCBI's)

In [480]:
rohit_calls: List[str] = list(data2['sequence'])
ncbi_calls: List[str] = list(data2['ncbi_sequence'])
[v for v in ncbi_calls][:2]

['MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQMEKIAEKKPQLLEKRMNMGFQKLADQLQESTK',
 'MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKSPKATKQMEKIAEKKPQLLEKRMNMGFQKLADQLQESTK']

### Iterates over ncbi/rohit calls, checking if each sequence is a subset of the other, or if they are identical

In [481]:
ncbi_subset_of_rohit: List[bool] = [v in rohit_calls[i] for i, v in enumerate(ncbi_calls)]
rohit_subset_of_ncbi: List[bool] = [v in ncbi_calls[i] for i, v in enumerate(rohit_calls)]
identical: List[bool] = [v == ncbi_calls[i] for i, v in enumerate(rohit_calls)]
print(' ncbi seqs contained in rohit seqs:', sum(ncbi_subset_of_rohit), 
      '\n', 'rohit seqs contained in ncbi seqs:', sum(rohit_subset_of_ncbi), 
      '\n', 'sequence calls are the same:', sum(identical),
      '\n', 'total seqs in data:', data2.shape[0])

 ncbi seqs contained in rohit seqs: 57 
 rohit seqs contained in ncbi seqs: 10 
 sequence calls are the same: 0 
 total seqs in data: 109


### Adding these subset seqs to a df

In [482]:
data2['ncbi_subset_of_rohit'] = ncbi_subset_of_rohit
data2['rohit_subset_of_ncbi'] = rohit_subset_of_ncbi
print(data2['sequence'].loc[4], data2['ncbi_sequence'].loc[4])

MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF


# Functions

### Finds insertions/deletions between two strings
- **Only true for one row at the beginning of the data. so insertions/deletions were not as common as it first appeared..**

In [483]:
def find_difference_two_strings(str1: str, str2: str):
    from difflib import ndiff
    return [li for li in ndiff(str1, str2) if li[0] != ' ']

In [484]:
row4_rohit = data2['sequence'].loc[4]
row4_ncbi = data2['ncbi_sequence'].loc[4]

print(' rohit:', row4_rohit, '\n', 'ncbi: ', row4_ncbi, '\n difference:', find_difference_two_strings(row4_rohit, row4_ncbi))

 rohit: MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF 
 ncbi:  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF 
 difference: ['+ I']


*May need this code to find the index where the insertions/deletions occur at as well.*

In [485]:
# Attempting to find the index of the difference between two strings
from difflib import ndiff, SequenceMatcher
test_a = 'MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF'
test_b = 'MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF'

display(SequenceMatcher(None, test_a, test_b).get_matching_blocks(), test_b[69])

[Match(a=0, b=0, size=69),
 Match(a=69, b=70, size=44),
 Match(a=113, b=114, size=0)]

'I'

In [486]:
def check_if_begin_end_same(seq1: str, seq2:str) -> bool:
    '''
    Tells you if the first three letters and last three chars of a string are both the same. 
    If not, an insertion, deletion or frameshift is likely.
    '''
    if (seq1.startswith(seq2[:3]) and seq1.endswith(seq2[-3:])):
        return True
    else:
        return False

In [487]:
# Running function on all spanin sequence data
data2['begin_end_same'] = [check_if_begin_end_same(rohit_calls[i], ncbi_calls[i]) for i, v in enumerate(rohit_calls)]
display(data2[data2['begin_end_same'] == True])

phage_acc    strand           cds  \
4  NC_016562.1  NEGATIVE  89202..89547   

                                            sequence  \
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   

                                       ncbi_sequence  ncbi_subset_of_rohit  \
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...                 False   

   rohit_subset_of_ncbi  begin_end_same  
4                 False            True

### Find the *length* difference between two strings, then based on which string is a subset of one another, determine the new cds
* Need to parse sequence and ncbi_sequence against one another, checking if one is a subset of another.
* If so, determine the difference in lengths between the larger one and the smaller
* Multiply this difference by three. The starting CDS should be approximately that many bases before or after the NCBI CDS (or +1 the NCBI CDS due to messed up indexing?)

In [488]:
def length_difference_two_strings(str1: str, str2: str) -> int:
    return abs(len(str1) - len(str2))

In [489]:
def check_if_subset(str1: str, str2: str) -> bool:
    return str1 in str2

In [490]:
def convert_ncbi_cds(cds: str) -> List[int]:
    """
    Splits ncbi cds data (Ex: '123..6789') into a list of two integers
    """
    return cds.split('..')

In [491]:
def find_new_cds(seq1_cds: str, seq1: str, seq2: str, strand: str) -> List[int]:
    """
    Finds a new cds given a previous cds, the old protein sequence, and a new protein sequence IFF one is a subset of the other.
    This assumes the annotations are in the same reading frame.
    It also assumes seq2's annotations are correct and the seq1 ones are not.
    """
    # Converts ncbi cds
    cds: List[str] = seq1_cds.split('..')
    cds: List[int] = [int(x) for x in cds]
    
    # Finds difference in length between two seqs
    length_difference: int = length_difference_two_strings(seq1, seq2)
    
    # Determines which is a subset of the other
    seq1_subset_of_seq2: bool = check_if_subset(seq1, seq2)
    seq2_subset_of_seq1: bool = check_if_subset(seq2, seq1)
    
    # Checks if they begin the same or end the same (first and last 3 characters)
    begin_same: bool = seq1.startswith(seq2[:3])
    end_same: bool = seq1.endswith(seq2[-3:])
    
    # Determines how much to shift the cds
    # Multiplies by 3, because cds indexes by nucleotides while passed seqs are protein seqs
    cds_shift: int = length_difference*3
    
    if strand == 'POSITIVE':
        # seq1 = ncbi_seq and seq2 = rohit_seq
        if seq1_subset_of_seq2:
            if begin_same:
                # extends seq1 cds end to seq2 cds end
                cds[1] = cds[1] + cds_shift
                return cds

            if end_same:
                # extends seq1 cds start to seq2 cds start
                cds[0] = cds[0] - cds_shift
                return cds

        elif seq2_subset_of_seq1:
            # reduces seq1 cds end to the seq2 cds end
            if begin_same:
                cds[1] = cds[1] - cds_shift 
                return cds

            if end_same:
                # reduces seq1 cds to match seq2 cds start
                cds[0] = cds[0] + cds_shift
                return cds
            
    if strand == 'NEGATIVE':
    # Flipped the signs for all. Need to verify that is correct.
        if seq1_subset_of_seq2:
            if begin_same:
                cds[1] = cds[1] - cds_shift
                return cds

            if end_same:
                cds[1] = cds[1] + cds_shift
                return cds

        elif seq2_subset_of_seq1:
            if begin_same:
                cds[0] = cds[0] + cds_shift 
                return cds

            if end_same:
                cds[1] = cds[1] - cds_shift
                return cds
        
    if begin_same:
        if end_same:
            # Imperfectly checks if there are insertions or deletions
            # Probably not worth writing the logic to handle this. Only one case in the data.
            return [1,1]
    
        # Probably need entirely different logic to handle this case
    if (not begin_same and not end_same):
        return [0,0]
    
    else:
        # these cases do not match because the ncbi_seq uses an uncommon start codon upstream of the one rohit selected (V, L or I mostly it looks like).
        # Handling these by finding the difference in length between ncbi_sequence and sequence, and changing the new cds to match rohit's sequence
        if strand == 'POSITIVE':
            if end_same:
                cds[0] = cds[0] - cds_shift 
                return cds
        if strand == 'NEGATIVE':
            if end_same:
                cds[1] = cds[1] + cds_shift
                return cds
    
    # failsafe to avoid returning None
    return [2,2]

## Applying the function

In [492]:
data2['new_cds'] = data2.apply(lambda x: find_new_cds(x['cds'], x['ncbi_sequence'], x['sequence'], x['strand']), axis=1)
data2.head(7)

phage_acc    strand             cds  \
0  NC_020879.1  NEGATIVE  126555..126789   
1  NC_019543.1  NEGATIVE  124571..124826   
3  NC_024142.1  POSITIVE    12841..13066   
4  NC_016562.1  NEGATIVE    89202..89547   
6  NC_007805.1  POSITIVE    37910..38360   
7   KC139515.1  NEGATIVE    17634..17832   
8  NC_010392.1  NEGATIVE    30405..30858   

                                            sequence  \
0  MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1  MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
3  MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   
6  MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   
7  MTGLLARIKTGVLAALVFVVALFGVWRAGRTKGKQDQINNQNNDTL...   
8  MMFNWKTMFVGLLLVSLIVAGRLANHYRNNAITYKYQRDTATHNLK...   

                                       ncbi_sequence  ncbi_subset_of_rohit  \
0  MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...                  True   
1  MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...                  True   
3  MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...                  True   
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...                 False   
6  MVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQ...                 False   
7  MALFGVWRAGRTKGKQDQINNQNNDTLREQANADKNVAEVHNEINK...                 False   
8  MFVGLLLVSLIVAGRLANHYRNNAITYKYQRDTATHNLKLANETIT...                  True   

   rohit_subset_of_ncbi  begin_end_same           new_cds  
0                 False           False  [126555, 126873]  
1                 False           False  [124571, 124889]  
3                 False           False    [12559, 13066]  
4                 False            True            [1, 1]  
6                 False           False    [37889, 38360]  
7                 False           False    [17634, 17889]  
8                 False           False    [30405, 30879]

In [493]:
data2[data2['new_cds'].isnull()]

Empty DataFrame
Columns: [phage_acc, strand, cds, sequence, ncbi_sequence, ncbi_subset_of_rohit, rohit_subset_of_ncbi, begin_end_same, new_cds]
Index: []

In [494]:
data.loc[62:66]

host  host_taxid    phage_acc     protein_acc  protein_gi  \
62  Pseudomonas         286  NC_020203.1  YP_007392789.1   448245071   
63  Pseudomonas         286   JN811560.1      AEY99484.1   374112688   
64  Pseudomonas         286  NC_020198.1  YP_007392334.1   448244611   
65  Pseudomonas         286  NC_020202.1  YP_007392731.1   448245012   
66  Pseudomonas         286  NC_020200.1  YP_007392432.1   448244711   

             cds    strand function spanin_type property_feature  \
62  12285..12399  POSITIVE  OSPANIN         ovl      LIPO:17..20   
63  12561..12675  POSITIVE  OSPANIN         ovl      LIPO:17..20   
64  12392..12506  POSITIVE  OSPANIN         ovl      LIPO:17..20   
65  12344..12458  POSITIVE  OSPANIN         ovl      LIPO:17..20   
66  11426..11540  POSITIVE  OSPANIN         ovl      LIPO:17..20   

                                             sequence  \
62  MPWPKPLLIALPAAFLLASCSSSKPPVNVPPRPLPAALAQPCPTPV...   
63  MPWPKPLLIALPAAFLLASCSSSKPPVNVPPRPLPAALAQPCPTPV...   
64  MPWPKPLLIALPAAFLLASCSSSKPPVNVPPRPLPAALAQPCPTPV...   
65  MPWPKPLLIALPAAFLLASCSSSKPPVNVPPRPLPAALAQPCQTPV...   
66  MPWPKPLLIALPAAFLLASCSSSKPPVNVPPRPLPAALAQPCPTPV...   

                            ncbi_sequence same  
62  MTDDSPDATAIALKQLYDQYGVCAGLHWDTVRHLQKD    N  
63  MTDDSPDATAIALKQLYDQYGLCAGLHWDTVRHLQKD    N  
64  MTDDSPDATAIALKQLYDQYGLCAGLHWDTVRHLQKD    N  
65  MTDDSPDAAAIALKQLYDQYGACAGLHWDTVRHLQKD    N  
66  MTDDSPDAAAIALKQLYDQYGACAGLHWDTVRHFQKD    N

In [495]:
list(data2['cds'].apply(lambda x: x.split('..')))

[['126555', '126789'],
 ['124571', '124826'],
 ['12841', '13066'],
 ['89202', '89547'],
 ['37910', '38360'],
 ['17634', '17832'],
 ['30405', '30858'],
 ['17493', '17973'],
 ['132794', '133061'],
 ['59169', '59394'],
 ['26697', '26961'],
 ['23912', '24308'],
 ['11928', '12216'],
 ['36334', '36736'],
 ['6803', '7103'],
 ['20330', '20630'],
 ['41067', '41313'],
 ['22226', '22775'],
 ['32853', '33276'],
 ['42438', '42759'],
 ['23635', '23845'],
 ['167032', '167455'],
 ['24156', '24534'],
 ['134824', '135121'],
 ['11233', '11458'],
 ['18144', '18336'],
 ['46829', '47042'],
 ['129054', '129321'],
 ['43580', '43853'],
 ['9447', '9939'],
 ['137726', '138050'],
 ['136877', '137234'],
 ['58896', '59115'],
 ['28053', '28608'],
 ['132260', '132545'],
 ['234503', '234857'],
 ['23239', '23539'],
 ['210927', '211197'],
 ['132642', '132858'],
 ['35414', '35684'],
 ['9381', '9528'],
 ['16184', '16514'],
 ['79273', '79471'],
 ['36139', '36415'],
 ['4906', '5065'],
 ['115946', '116180'],
 ['5604', '6030'

In [496]:
from Bio import Entrez
from Bio import SeqIO


class GetPhageSeq:
    """
    Connnects to the NUCCORE database and retrieves phage sequence.
        Email : Email used by NCBI to notifiy with errors and problems
        acc : Accession use to query database
        db : database
    """
    def __init__(self,email,acc,db):
        self.email = email
        self.acc = acc
        self.db = db
        Entrez.email = self.email
    
    def get_and_return_sequence(self, ret_type):
        """
        Uses input params to retrieve the sequence from NCBI
        """

        sequence = Entrez.efetch(db=self.db, id=self.acc, rettype=ret_type, retmode="text")
        read_fasta = str(SeqIO.read(sequence,"fasta").seq)

        return read_fasta



# if __name__ == '__main__':

#     payload = {
#         "email" : "curtisross@tamu.edu",
#         "acc" : "NC_001416.1",
#         "db" : "nuccore"
#     }

    # gps = GetPhageSeq(**payload).get_and_return_sequence(ret_type="fasta")
    #sequence = str(SeqIO.read(gps,"fasta").seq)


    # print(gps)
    # print(len(gps))
    # print(sequence)

In [497]:
payload = {
        "email" : "curtisross@tamu.edu",
        "acc" : "NC_020879.1",
        "db" : "nuccore"
    }
from Bio.Seq import Seq

phage1 = Seq(GetPhageSeq(**payload).get_and_return_sequence(ret_type="fasta"))


# Phage "NC_007805.1" 

In [498]:
payload3 = {
        "email" : "curtisross@tamu.edu",
        "acc" : "NC_007805.1",
        "db" : "nuccore"
    }
phage3pos_strand = Seq(GetPhageSeq(**payload3).get_and_return_sequence(ret_type="fasta"))

In [499]:
str(phage3pos_strand[37910:38360].translate(to_stop=True))

'VVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQAQVQTDALARTYQAALQASHEENQLRRDAIGTGARVVYVKARCPAGGVHQAPGATGSADAGRAVLAAADGQVVSDLRAGVERRELMIAALRKHIAGLPRYCRR'

In [500]:
str(phage3pos_strand[37889:38360].translate(to_stop=True))

'MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQAQVQTDALARTYQAALQASHEENQLRRDAIGTGARVVYVKARCPAGGVHQAPGATGSADAGRAVLAAADGQVVSDLRAGVERRELMIAALRKHIAGLPRYCRR'

In [501]:
data2

phage_acc    strand             cds  \
0    NC_020879.1  NEGATIVE  126555..126789   
1    NC_019543.1  NEGATIVE  124571..124826   
3    NC_024142.1  POSITIVE    12841..13066   
4    NC_016562.1  NEGATIVE    89202..89547   
6    NC_007805.1  POSITIVE    37910..38360   
..           ...       ...             ...   
116  NC_021794.1  POSITIVE    41131..41452   
117  NC_021789.1  POSITIVE    44760..45078   
119  NC_019540.1  NEGATIVE    20740..21175   
120  NC_024791.1  NEGATIVE        629..968   
121  NC_007603.1  POSITIVE    41476..41839   

                                              sequence  \
0    MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1    MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
3    MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   
4    MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   
6    MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   
..                                                 ...   
116  MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNI...   
117  MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNI...   
119  MKRLRQVLLMLGMVLALSGCSAVTSAVTDRLAGGDKPAVGIDTEIV...   
120  MTYVRNKMKSLLVLMGLVLALSTNGCSSLTPVEIAKDVLLPDQQSG...   
121  MLKWIAALVMCVALSGCSATSALTGLIGSKPEISAQAGAENNKVGV...   

                                         ncbi_sequence  ncbi_subset_of_rohit  \
0    MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...                  True   
1    MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...                  True   
3    MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...                  True   
4    MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...                 False   
6    MVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQ...                 False   
..                                                 ...                   ...   
116  MMLFSIACFAQDKNYVSVHRDSLYNALLNISELKVENAFLRKKNDN...                  True   
117  MLFSIACFAQDKNYVSVHRDSLYNALLNISELKVENAFLRKKNDNQ...                  True   
119  MLSRLCQKHLDLQGTTMKRLRQVLLMLGMVLALSGCSAVTSAVTDR...                 False   
120  MGLVLALSTNGCSSLTPVEIAKDVLLPDQQSGITVDTQIGDKEYAL...                  True   
121  MCVALSGCSATSALTGLIGSKPEISAQAGAENNKVGVGLTNKVDSS...                  True   

     rohit_subset_of_ncbi  begin_end_same           new_cds  
0                   False           False  [126555, 126873]  
1                   False           False  [124571, 124889]  
3                   False           False    [12559, 13066]  
4                   False            True            [1, 1]  
6                   False           False    [37889, 38360]  
..                    ...             ...               ...  
116                 False           False    [41083, 41452]  
117                 False           False    [44709, 45078]  
119                  True           False    [20740, 21127]  
120                 False           False       [629, 1010]  
121                 False           False    [41449, 41839]  

[109 rows x 9 columns]

In [502]:
def test_phage_sequence(acc: str, cds: str, new: List[int], strand: str, rohitseq: str):
           
    payload = {
        "email" : "curtisross@tamu.edu",
        "acc" : f"{acc}",
        "db" : "nuccore"
    }
    
    # Imports genome as FASTA
    genome = Seq(GetPhageSeq(**payload).get_and_return_sequence(ret_type="fasta"))
    
    # Converts ncbi cds
    old_cds: List[str] = cds.split('..')
    old_cds: List[int] = [int(x) for x in old_cds]
        
    start: int = old_cds[0]
    end: int = old_cds[1]
    
    if strand == 'POSITIVE':
        if rohitseq == str(genome[new[0]:new[1]].translate(to_stop=True)):
            return True
        else:
            return str(find_difference_two_strings(rohitseq, str(genome[new[0]:new[1]].translate(to_stop=True))))
        
    if strand == 'NEGATIVE':
        if rohitseq == str(genome[new[0]:new[1]].reverse_complement().translate(to_stop=True)):
            return True
        else:
            # return rohitseq + '   ' + str(genome[new_cds[0]:new_cds[1]].reverse_complement().translate(to_stop=True))
            #return str(genome[126471:126789].reverse_complement().translate(to_stop=True))
            #return rohitseq == str(genome[new_cds[0]:new_cds[1]].reverse_complement().translate(to_stop=True))
            return find_difference_two_strings(rohitseq, str(genome[new[0]:new[1]].reverse_complement().translate(to_stop=True)))

In [503]:
data2['verified'] = data2.apply(lambda x: test_phage_sequence(x['phage_acc'], x['cds'], x['new_cds'], x['strand'], x['sequence']), axis=1)

In [504]:
data2.head()

phage_acc    strand             cds  \
0  NC_020879.1  NEGATIVE  126555..126789   
1  NC_019543.1  NEGATIVE  124571..124826   
3  NC_024142.1  POSITIVE    12841..13066   
4  NC_016562.1  NEGATIVE    89202..89547   
6  NC_007805.1  POSITIVE    37910..38360   

                                            sequence  \
0  MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1  MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
3  MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   
6  MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   

                                       ncbi_sequence  ncbi_subset_of_rohit  \
0  MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...                  True   
1  MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...                  True   
3  MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...                  True   
4  MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...                 False   
6  MVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQ...                 False   

   rohit_subset_of_ncbi  begin_end_same           new_cds  \
0                 False           False  [126555, 126873]   
1                 False           False  [124571, 124889]   
3                 False           False    [12559, 13066]   
4                 False            True            [1, 1]   
6                 False           False    [37889, 38360]   

                                            verified  
0                                               True  
1                                               True  
3  ['+ G', '+ I', '- K', '- M', '- L', '+ A', '+ ...  
4  [- M, - Q, - M, - F, - N, - F, - L, - F, - S, ...  
6                                               True

In [505]:
true = data2[data2['verified'] == True]
true

phage_acc    strand             cds  \
0    NC_020879.1  NEGATIVE  126555..126789   
1    NC_019543.1  NEGATIVE  124571..124826   
6    NC_007805.1  POSITIVE    37910..38360   
7     KC139515.1  NEGATIVE    17634..17832   
8    NC_010392.1  NEGATIVE    30405..30858   
..           ...       ...             ...   
116  NC_021794.1  POSITIVE    41131..41452   
117  NC_021789.1  POSITIVE    44760..45078   
119  NC_019540.1  NEGATIVE    20740..21175   
120  NC_024791.1  NEGATIVE        629..968   
121  NC_007603.1  POSITIVE    41476..41839   

                                              sequence  \
0    MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1    MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
6    MRWVPWLVVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQ...   
7    MTGLLARIKTGVLAALVFVVALFGVWRAGRTKGKQDQINNQNNDTL...   
8    MMFNWKTMFVGLLLVSLIVAGRLANHYRNNAITYKYQRDTATHNLK...   
..                                                 ...   
116  MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNI...   
117  MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNI...   
119  MKRLRQVLLMLGMVLALSGCSAVTSAVTDRLAGGDKPAVGIDTEIV...   
120  MTYVRNKMKSLLVLMGLVLALSTNGCSSLTPVEIAKDVLLPDQQSG...   
121  MLKWIAALVMCVALSGCSATSALTGLIGSKPEISAQAGAENNKVGV...   

                                         ncbi_sequence  ncbi_subset_of_rohit  \
0    MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...                  True   
1    MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...                  True   
6    MVALVAALVFWRLDHVTAQRNDLQAAVEQSAETITAMAQQAQRDTQ...                 False   
7    MALFGVWRAGRTKGKQDQINNQNNDTLREQANADKNVAEVHNEINK...                 False   
8    MFVGLLLVSLIVAGRLANHYRNNAITYKYQRDTATHNLKLANETIT...                  True   
..                                                 ...                   ...   
116  MMLFSIACFAQDKNYVSVHRDSLYNALLNISELKVENAFLRKKNDN...                  True   
117  MLFSIACFAQDKNYVSVHRDSLYNALLNISELKVENAFLRKKNDNQ...                  True   
119  MLSRLCQKHLDLQGTTMKRLRQVLLMLGMVLALSGCSAVTSAVTDR...                 False   
120  MGLVLALSTNGCSSLTPVEIAKDVLLPDQQSGITVDTQIGDKEYAL...                  True   
121  MCVALSGCSATSALTGLIGSKPEISAQAGAENNKVGVGLTNKVDSS...                  True   

     rohit_subset_of_ncbi  begin_end_same           new_cds verified  
0                   False           False  [126555, 126873]     True  
1                   False           False  [124571, 124889]     True  
6                   False           False    [37889, 38360]     True  
7                   False           False    [17634, 17889]     True  
8                   False           False    [30405, 30879]     True  
..                    ...             ...               ...      ...  
116                 False           False    [41083, 41452]     True  
117                 False           False    [44709, 45078]     True  
119                  True           False    [20740, 21127]     True  
120                 False           False       [629, 1010]     True  
121                 False           False    [41449, 41839]     True  

[85 rows x 10 columns]

In [506]:
false = data2[data2['verified'] != True]
false

phage_acc    strand             cds  \
3    NC_024142.1  POSITIVE    12841..13066   
4    NC_016562.1  NEGATIVE    89202..89547   
14   NC_005884.1  POSITIVE    11928..12216   
16    GQ422450.1  POSITIVE      6803..7103   
17    GQ422451.1  POSITIVE    20330..20630   
21   NC_019926.1  POSITIVE    42438..42759   
23     NC_016163  NEGATIVE  167032..167455   
28      AM076770  POSITIVE    18144..18336   
34   NC_024794.1  NEGATIVE  136877..137234   
38   NC_023568.1  POSITIVE  234503..234857   
44   NC_014662.1  NEGATIVE  132642..132858   
46   NC_023576.1  POSITIVE    35414..35684   
50     NC_011042  POSITIVE    36139..36415   
69   NC_007456.1  POSITIVE    36645..36921   
72   NC_013647.1  POSITIVE    38218..38710   
84   NC_019528.1  NEGATIVE    13794..13980   
87   NC_006938.1  POSITIVE       903..1020   
90      AM076770  POSITIVE    18310..18466   
94   NC_013597.1  NEGATIVE    23051..23264   
95     NC_016763  POSITIVE    23651..23795   
98   NC_022749.1  POSITIVE    38716..39058   
100   JQ177063.1  NEGATIVE          0..117   
109  NC_019713.1  POSITIVE    37640..37760   
115  NC_021803.1  POSITIVE    14606..14918   

                                              sequence  \
3    MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQK...   
4    MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...   
14   MYKIAIGLALGLALFSLWSYYSLQLTKQELATTKEALADRSKEVEF...   
16   MSRIKAIIASVIICIIVCLSWAVNHYRDNAITYKEQRDKATSIIAD...   
17   MSRIKAIIASVIICIIVCLSWAVNHYRDNAITYKEQRDKATSIIAD...   
21   MRKIYVVIITTIVMAGLIWAFIATQVNTGVTSKRQEDALAVSEANV...   
23   MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWE...   
28   MLGKLKIAVMLMIAAVLAWKAGSWNGARVERSVQIAECNNRIEKLA...   
34   MLKVNPIYAIVAAFALVSTVTIVVLNNKVDSLNTELASVKETAKNN...   
38   MFSQNKIYVLLAVAAVFIGSIGYLKYENVSLEKDLVTQTAEVERLT...   
44   MKLLVATVLSIGLLAGCSQKTPELPQPTVVHPSWPDQIKPWNGHWT...   
46   MLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQ...   
50   MLSKLKSTKCPNGFRTKCHRWKVALIGLLLTFSLTISGCASESKLR...   
69   MLSKLKSTKCPNGFRTKCPRWKAALIGLLLTLSLTISGCSSESNLR...   
72   MTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQ...   
84   MKLLNRVWSVAMLWLIAGLLMSCATQHVAPKATPMHVPEALTAPTP...   
87   MTLSQEPVRSLLRGLSIVALLVLGACSSVPPERSVIQSVQVKTPKT...   
90   MLAKLNKAFWLLLIAILISGCQSGRVITVPVVVDTSCDWVEVIRPT...   
94   MKMRKPLKQSLKLNLVIVLSSLSLLLSGCTSKVTTKAEYIYPPQAY...   
95   MKLSNTLPVLAVLSVSFLLNGCQSNNAPLRMLIPSPDMTLMQPPCK...   
98   MRMMIVLYALCLLPLLTGCESSRTVYVPVPAIPLPASLTAETPQPA...   
100  MNGSKTPLCIAALLIAGCSSSSTPPTALIAPSADLLVPPGEPLPVC...   
109  MRLVKLTLSSLFLVMSMSACTTTQYVQPTEPPKKDIPPELLKRCPP...   
115  MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNI...   

                                         ncbi_sequence  ncbi_subset_of_rohit  \
3    MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEEL...                  True   
4    MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...                 False   
14   MATTKEALADRSKEVEFLNTSLQLRDKVALQTAEREGAINAQLQRI...                 False   
16   MQKRQRDVAELDARYTKELADANATIETLRADVSAGRKRLQVSATC...                  True   
17   MQKRQRDVAELDARYTKELADANATIETLRADVSAGRKRLQVSATC...                  True   
21   MAGLIWAFIATQVNTGVTSKRQEDALAVSEANVGIGKEAKDQGEQA...                  True   
23   MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWE...                 False   
28   MQIAECNNRIEKLAAELEAEKAKKKVEVTKSASKTKQSVLVATDSD...                 False   
34   MLKVNPIYAIVAAFALVSTVTIVVLNNKVDSLNTELASVKETAKNN...                 False   
38   MAVAAVFIGSIGYLKYENVSLEKDLVTQTAEVERLTGDNERLQTTI...                 False   
44   MHPSWPDQIKPWNGHWTIVEVDGRPFVGMPYSDSQEFRIWLNDVLR...                 False   
46   MQNTPRILRHRKIQGLLSKLKSTKCPSGFRTKCPRWKAALIGLLLT...                 False   
50   MQNTPRILRHRRIQGLLSKLKSTKCPNGFRTKCHRWKVALIGLLLT...                 False   
69   MRNTPRILRHRKIQGLLSKLKSTKCPNGFRTKCPRWKAALIGLLLT...                 False   
72   MTISSPYSDILKVLLYVVPLWMSLHYARRSYVKSEKHLEALCWKLQ...                 False   
84   MHVPEALTAPTPAPEIEPGIRISDYPKLILRSLGALEEANADKAAV...                  True   
87              MALTPRQYQNLSNNLAEVLRYVKQADSQL

### Phage 3   = Phage 3 is a mess. Genome was updated to NC_024142.2. New cds in ncbi is 12842..13066, however their CDS in ncbi is not a multiple of three. AND I get a different sequence when I translate that cds (probably because the genome in NCBI is a slightly different size than the one efetch gets me...?)
### Phage 23 = Rohit's annotation appears to be incorrect here. He ends one amino acid before the stop codon, while NCBI's annotation correctly goes to the stop codon
### Phage 34 = 137084:137234, NC_024794.1 (manually)
### Phage 46 =  35459:35684, NC_023576.1
### Phage 50 =  36184:36415
### Phage 69 = This one is a mess too. The protein_acc and protein_gi are out of date and do not match anything in ncbi. They appear to have re-annotated it  by shifting the cds from 36645:36921 to 36550:37000 - which is completely different from rohit's annotation and their original annotation - not even in the same reading frame
### Phage 72 = 38503:38710 **but** I find rohit's annotation a little questionable here. He's missing a huge part of the orf
### Phage 98 = 38776:39058
### Phage 115 = My guess is that they annotated different genes here because the sequence is completely different

In [507]:
payload46 = {
        "email" : "curtisross@tamu.edu",
        "acc" : "NC_023576.1",
        "db" : "nuccore"
    }
from Bio.Seq import Seq

phage46 = Seq(GetPhageSeq(**payload46).get_and_return_sequence(ret_type="fasta"))
# negative strand so need to rev_complement

print('', phage46[35414:35684].translate(), '\n', phage46[35369:35684].translate()),'\n', 'length:', len(phage46)

 MQNTPRILRHRKIQGLLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE* 
 LVATTPLTISGRLRSMQNTPRILRHRKIQGLLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE*


(None, '\n', 'length:', 39207)

In [508]:
data2['sequence'][46], data2['ncbi_sequence'][46]

('MLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE',
 'MQNTPRILRHRKIQGLLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE')

In [509]:
def test_spanin(acc: str, cds: List[int], test_cds: List[int], strand: str):
    payloadtest = {
        "email" : "curtisross@tamu.edu",
        "acc" : f"{acc}",
        "db" : "nuccore"
                }
    
    phage = Seq(GetPhageSeq(**payloadtest).get_and_return_sequence(ret_type="fasta"))
    
    if strand == 'NEGATIVE':
        return print('', phage[cds[0]:cds[1]].reverse_complement().translate(), '\n', phage[test_cds[0]:test_cds[1]].reverse_complement().translate()),'\n', 'length:', len(phage)
    return print('', phage[cds[0]:cds[1]].translate(), '\n', phage[test_cds[0]:test_cds[1]].translate()),'\n', 'length:', len(phage)

In [510]:
# Had to update the acc to the updated version for phage3
phage3 = test_spanin('NC_024142.2', [12841, 13066], test_cds = [12842,13066], strand=data2['strand'][3])
phage3

 MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEELRATVRLRDSQLIELGKQIQADRKLFEQE* 
 CNSLKDEHRYINVKPKPEPLSAEVLQAMQPDSTTVLKKADVWLKNSGQLLDSVTAN*LSWVNRFRLIVNFLNRN


C:\Users\jcham\Anaconda3\lib\site-packages\Bio\Seq.py:2742: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
  BiopythonWarning,


(None, '\n', 'length:', 72583)

In [511]:
data2['sequence'][3], data2['ncbi_sequence'][3]

('MKMLISKGWPYLLVVVLGATIYFWGNSNGQSTVQKKWDDQKVEDQKAMQKLQDKYNALQRNHSYEVGLLTSRLQTAESNYASELARVSSDYDSRMQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEELRATVRLRDSQLIELGKQIQADRKLFEQE',
 'MQQSERRASVYKRQAEAGTFECRSLASHAARLDNSLEEGRRLVEELRATVRLRDSQLIELGKQIQADRKLFEQE')

#### Rohit's annotation stops one amino acid before the stop codon here. NCBI annotation is the correct one

In [512]:
phage23 = test_spanin(data2['phage_acc'][23], [167032, 167455], test_cds = [167035,167455], strand=data2['strand'][23])
phage23

 MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWEGKYTAQKLVSDQLEAALKAEQDKNKKLEAANAIQQTEIANLKAEDNKLNDIDKSNVEVIIKEVPADKVISKESSKALVKIADNFDTVVDNLEDL* 
 MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWEGKYTAQKLVSDQLEAALKAEQDKNKKLEAANAIQQTEIANLKAEDNKLNDIDKSNVEVIIKEVPADKVISKESSKALVKIADNFDTVVDNLEDL


(None, '\n', 'length:', 262391)

In [513]:
display('rohit', data2['sequence'][23], 'ncbi', data2['ncbi_sequence'][23])

'rohit'

'MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWEGKYTAQKLVSDQLEAALKAEQDKNKKLEAANAIQQTEIANLKAEDNKLNDIDKSNVEVIIKEVPADKVISKESSKALVKIADNFDTVVDNLED'

'ncbi'

'MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWEGKYTAQKLVSDQLEAALKAEQDKNKKLEAANAIQQTEIANLKAEDNKLNDIDKSNVEVIIKEVPADKVISKESSKALVKIADNFDTVVDNLEDL'

In [514]:
data2['sequence'][23]

'MDSILNNFVSSYKNIILAICIAIVVVVIGLFLNGIADNAKASTEWEGKYTAQKLVSDQLEAALKAEQDKNKKLEAANAIQQTEIANLKAEDNKLNDIDKSNVEVIIKEVPADKVISKESSKALVKIADNFDTVVDNLED'

In [515]:
phage46 = test_spanin(data2['phage_acc'][46], [35414, 35684], test_cds = [35459,35684], strand=data2['strand'][46])
phage46

 MQNTPRILRHRKIQGLLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE* 
 LLSKLKSTKCPSGFRTKCPRWKAALIGLLLTLTAITSGCASKSTPQVSPSQISVDASLMVESNYTQKLLKVLSE*


(None, '\n', 'length:', 39207)

In [516]:
phage50 = test_spanin(data2['phage_acc'][50], [36139, 36415], test_cds = [36184,36415], strand=data2['strand'][50])
phage50

 MQNTPRILRHRRIQGLLSKLKSTKCPNGFRTKCHRWKVALIGLLLTFSLTISGCASESKLRVEPRKVTVDASLMVTPNLTNEMLNVLSPSE* 
 LLSKLKSTKCPNGFRTKCHRWKVALIGLLLTFSLTISGCASESKLRVEPRKVTVDASLMVTPNLTNEMLNVLSPSE*


(None, '\n', 'length:', 39252)

In [517]:
phage69 = test_spanin(data2['phage_acc'][69], [36550, 37000], test_cds = [36550,37000], strand=data2['strand'][69])
phage69

 MLEFLKRAAPWLLAAVMFAGGYHTANNKWEAKVNAEYTSNLKASEDTRLAVQAEVNKVSKRFQDEMSSLEGSTDRIIADLKSDNKRLLIRVKPTSGTTQSDGRCFIDGYAELDERDAKRLIAIGVKGDKWIKALQDTVRALQQEKEVTH* 
 MLEFLKRAAPWLLAAVMFAGGYHTANNKWEAKVNAEYTSNLKASEDTRLAVQAEVNKVSKRFQDEMSSLEGSTDRIIADLKSDNKRLLIRVKPTSGTTQSDGRCFIDGYAELDERDAKRLIAIGVKGDKWIKALQDTVRALQQEKEVTH*


(None, '\n', 'length:', 39704)

In [518]:
data['sequence'][69], data['ncbi_sequence'][69]

('MLSKLKSTKCPNGFRTKCPRWKAALIGLLLTLSLTISGCSSESNLRVEPRKVTVDASLMVTPNLTNEMLNVLSPSE',
 'MRNTPRILRHRKIQGLLSKLKSTKCPNGFRTKCPRWKAALIGLLLTLSLTISGCSSESNLRVEPRKVTVDASLMVTPNLTNEMLNVLSPSE')

In [519]:
data.iloc[69]

host                                                   Enterobacteria
host_taxid                                                        547
phage_acc                                                 NC_007456.1
protein_acc                                                CAJ29394.1
protein_gi                                                   83308161
cds                                                      36645..36921
strand                                                       POSITIVE
function                                                      OSPANIN
spanin_type                                                       emb
property_feature                                          LIPO:36..39
sequence            MLSKLKSTKCPNGFRTKCPRWKAALIGLLLTLSLTISGCSSESNLR...
ncbi_sequence       MRNTPRILRHRKIQGLLSKLKSTKCPNGFRTKCPRWKAALIGLLLT...
same                                                                N
Name: 69, dtype: object

In [520]:
phage72 = test_spanin(data2['phage_acc'][72], [38218, 38710], test_cds = [38218,38710], strand=data2['strand'][72])
phage72

 MTISSPYSDILKVLLYVVPLWMSLHYARRSYVKSEKHLEALCWKLQRESSHILWLSWCSPSGGTWVHNLRTPNGRRKYRMSTLRSKRLELKLRKRLTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ* 
 MTISSPYSDILKVLLYVVPLWMSLHYARRSYVKSEKHLEALCWKLQRESSHILWLSWCSPSGGTWVHNLRTPNGRRKYRMSTLRSKRLELKLRKRLTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ*


(None, '\n', 'length:', 41119)

#### Rohit is missing a huge part of the orf here compared to the ncbi annotation, and does not have a stop codon to justify it

In [521]:
data2['sequence'][72], data2['ncbi_sequence'][72]

('MTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ',
 'MTISSPYSDILKVLLYVVPLWMSLHYARRSYVKSEKHLEALCWKLQRESSHILWLSWCSPSGGTWVHNLRTPNGRRKYRMSTLRSKRLELKLRKRLTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ')

In [522]:
phage72 = test_spanin(data2['phage_acc'][72], [38218, 38710], test_cds = [38503,38710], strand=data2['strand'][72])
phage72

 MTISSPYSDILKVLLYVVPLWMSLHYARRSYVKSEKHLEALCWKLQRESSHILWLSWCSPSGGTWVHNLRTPNGRRKYRMSTLRSKRLELKLRKRLTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ* 
 LTQYRLSTKQTLRGWRAALIGLLLICVATISGCASKSNLPVSPQDQTVDASLMVSSNYTKQLLEVLSQ*


(None, '\n', 'length:', 41119)

In [523]:
phage98 = test_spanin(data2['phage_acc'][98], [38716, 39058], test_cds = [38776,39058], strand=data2['strand'][98])
phage98

 MQNNRSHWSHREPRLISKWLLRMMIVLYALCLLPLLTGCESSRTVYVPVPAIPLPASLTAETPQPAISEPLTYAGSLDLNVSLLSALGQCNLDKAGIRRIEASRSGRSESGSK* 
 LRMMIVLYALCLLPLLTGCESSRTVYVPVPAIPLPASLTAETPQPAISEPLTYAGSLDLNVSLLSALGQCNLDKAGIRRIEASRSGRSESGSK*


(None, '\n', 'length:', 39758)

In [524]:
phage115 = test_spanin(data2['phage_acc'][115], [14606, 14918], test_cds = [14606,14918], strand=data2['strand'][115])
phage115

 MNIYEFSYSGGETDWVFAPNIKEAKDFYLKFTGCGDLTLTTVKRVPKSKWNEMYLLDPNYSEPDEDEEDYNEEDYSCGLKIIETFAEYAERNTITDMIATTEY* 
 MNIYEFSYSGGETDWVFAPNIKEAKDFYLKFTGCGDLTLTTVKRVPKSKWNEMYLLDPNYSEPDEDEEDYNEEDYSCGLKIIETFAEYAERNTITDMIATTEY*


(None, '\n', 'length:', 72369)

In [525]:
data2['sequence'][115], data2['ncbi_sequence'][115]

('MEKTSNTSKNFMIVILMMLFSIACFAQDKNYVSVHRDSLYNALLNISELKVENAFLRKKNDNQKVIITAQEIIISETEFKTEILTKQISIEQSTGTKKWWNGFGKGVATGGVIVLGLFLGLK',
 'MNIYEFSYSGGETDWVFAPNIKEAKDFYLKFTGCGDLTLTTVKRVPKSKWNEMYLLDPNYSEPDEDEEDYNEEDYSCGLKIIETFAEYAERNTITDMIATTEY')

# NaN rows
#### 45, 79 and 99

## Phage 45 = Deletion of T in rohit's annotation compared to NCBI's annotation

In [526]:
phage45 = test_spanin(data['phage_acc'][45], [99401, 99680], test_cds = [99401,99680], strand=data['strand'][45])
phage45

 MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAETNYELLKIIKNKDLK* 
 MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAETNYELLKIIKNKDLK*


(None, '\n', 'length:', 175720)

In [527]:
data['sequence'][45], data['ncbi_sequence'][45]

('MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAENYELLKIIKNKDLK',
 'MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAETNYELLKIIKNKDLK')

In [528]:
find_difference_two_strings(
    'MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAENYELLKIIKNKDLK',
    'MKFKKFLLALIPFIFVGCSAVRTEFIYPKIPDVKEPPMTQDYNLTVIKINNVEYYSLSPEDAKILSENWIKFKSWAETNYELLKIIKNKDLK')

['+ T']

## Phage 79 = Genome has been updated to NC_021342.2
* They annotate the IM spanin (26746..27057) as embedded in the OM spanin (26592..26849)
* **Rohit only annotated the OM spanin cds here (26592:26849), which is the same cds they have.**

In [529]:
phage79 = test_spanin(data['phage_acc'][79], [26592, 26849], test_cds = [26592,26849], strand='NEGATIVE')
phage79

 MSIVSKLLLCGIVCAIAGCSNSTTRHSDCPTVSSDLLVPPRKLESTHNDPERAAWVIPWNAEALLADRSRLERWQAWYNATRKGL 
 MSIVSKLLLCGIVCAIAGCSNSTTRHSDCPTVSSDLLVPPRKLESTHNDPERAAWVIPWNAEALLADRSRLERWQAWYNATRKGL


(None, '\n', 'length:', 43418)

In [530]:
data['sequence'][79]

'MSIVSKLLLCGIVCAIAGCSNSTTRHSDCPTVSSDLLVPPRKLESTHNDPERAAWVIPWNAEALLADRSRLERWQAWYNATRKGL'

## Phage 99 = Genome has been updated to NC_012223.2
* They have updated the annotation and it is now the same as rohit's

In [531]:
phage99 = test_spanin(data['phage_acc'][99], [4835, 5066], test_cds = [4835,5066], strand='POSITIVE')
phage99

 MLRALTALIVVLTLSACGSSSVPLMPLTESALRPPCKFDNPSADPDEDLMIDVKNMECGAKLRAQVLELQQIIKGP* 
 MLRALTALIVVLTLSACGSSSVPLMPLTESALRPPCKFDNPSADPDEDLMIDVKNMECGAKLRAQVLELQQIIKGP*


(None, '\n', 'length:', 44899)

In [532]:
len('MLRALTALIVVLTLSACGSSSVPLMPLTESALRPPCKFDNPSADPDEDLMIDVKNMECGAKLRAQVLELQQIIKGP')

76

# Rows not called by NCBI submitter
#### 2, 5, 25, 40, 41, 42, 74, 81, 114, 118

### Phage 2 = called in NCBI now as hypothetical protein with CDS 134887:135240. Identical to rohit's sequence. Negative strand.
### Phage 5 = called in NCBI now as "Rz" with CDS 17544:18005. Identical to rohit's sequence. Negative strand.
### Phage 25 = called in NCBI now as hypothetical protein with CDS 41130:41510. Identical to rohit's sequence. However 41129:41510 is what properly translates in biopython. Pos strand.
### Phage 40 = called in NCBI now as hypothetical protein with CDS 5432:5779. Identical to rohit's sequence. However 5431:5779 is what properly translates in biopython. Pos strand.
### Phage 41 = called in NCBI now as 18.7 with CDS 35510:35761. Identical to rohit's sequence. However 35509:35761 is what properly translates in biopython. Pos strand.
### Phage 42 = called in NCBI as conserved hypothetical protein with CDS 134591:134890. Identical to rohit's sequence. Negative strand.
### Phage 74 = called in NCBI as hypothetical protein with CDS 54219:54575. Identical to rohit's sequence. 54218:54575 is what properly translates. Pos strand.
### Phage 81 = Rohit's call not found anywhere in the NCBI proteins. No spanins or hypothetical lipoproteins called in the NCBI data either. 
### Phage 114 = called in NCBI as hypothetical protein with CDS 5673:5942. Identical to rohit's sequence but 5672:5942 is what properly translates. Pos strand.
### Phage 118 = called in NCBI as hypothetical protein with CDS 56741:57115. Identical to rohit's sequence but 56740:57115 is what properly trnaslates. Pos strand.

In [533]:
phage2 = test_spanin(data['phage_acc'][2], [134887, 135240], test_cds = [134887,135240], strand='NEGATIVE')
phage2

 MHVSNFTAGLLLLVIAFGGTSIILKNKVERLETSVVEITKTANENALALNNLRIQYNYIDAMNNKNREAIAAIERENEKLRKDAKKADVVARKPGLVEKQINNSFNKFAEDIQDLSK 
 MHVSNFTAGLLLLVIAFGGTSIILKNKVERLETSVVEITKTANENALALNNLRIQYNYIDAMNNKNREAIAAIERENEKLRKDAKKADVVARKPGLVEKQINNSFNKFAEDIQDLSK


(None, '\n', 'length:', 167435)

In [534]:
phage4 = test_spanin(data['phage_acc'][4], [89203, 89547], test_cds = [89203,89547], strand='NEGATIVE')
phage4

 MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF 
 MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF


(None, '\n', 'length:', 132662)

In [535]:
data.iloc[4]

host                                                    Campylobacter
host_taxid                                                        194
phage_acc                                                 NC_016562.1
protein_acc                                            YP_004956969.1
protein_gi                                                  371671013
cds                                                      89202..89547
strand                                                       NEGATIVE
function                                                      ISPANIN
spanin_type                                                       ovl
property_feature                                           TMD:15..34
sequence            MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...
ncbi_sequence       MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEK...
same                                                                N
Name: 4, dtype: object

In [536]:
data['sequence'][4], data['ncbi_sequence'][4]

('MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF',
 'MQMFNFLFSFIKSNIIYILLGSLLAFTAYRYISLEKSNAILIENEKQLTQNIKNSKKELEALKNYNNLTIEVFREKEVKYKEVLNNIKNIETKIQKLKLMRKDENETQYIIVNF')

In [537]:
phage5 = test_spanin(data['phage_acc'][5], [134887, 135240], test_cds = [134887,135240], strand='NEGATIVE')
phage5

(None, '\n', 'length:', 42925)

In [538]:
data['sequence'][5]

'MSRVTAIISALVICIIVCLSWAVNHYRDNAITYKAQRDKNARELKLANAAITDMQMRQRDVAALDAKYTKELADAKAENDALRDDVAAGRRRLHIKAVCQSVREATTASGVDNAASPRLADTAERDYFTLRERLITMQKQLEGTQKYINEQCR'

In [539]:
phage25 = test_spanin(data['phage_acc'][25], [41129, 41510], test_cds = [41130,41510], strand='POSITIVE')
phage25

 LLTIPNKYKWAVMALLAAVSIGSLTLANHYRDSALTSQKALQEVTDAHVAYIEGVEAGEKTAGSRIEANKQLKERTDAAIKRVESIPGPTERVPDAVLIELQRAAEAASSSTRTSLPQKPRSSTRG* 
 CLLYRTSTNGL*WPC*PL*ASVV*PWLITTGIAH*PRRRLSRRSQMHMWLI*KVWKLVRRLLVPVSKLTSS*RREPMLQSRGWSLSLVPLSVFLMLSLSSCSVQPRPPVPVHVPASLRSPVAAPEV


(None, '\n', 'length:', 42085)

In [540]:
phage40 = test_spanin(data['phage_acc'][40], [5431, 5779], test_cds = [5432,5779], strand='POSITIVE')
phage40

 MTFKEFCQATFIIVFLVGAGVWGGYSYRDYQVAEAELSNEKLISVAKDAYQEGLVTLSTNYKNDLKDVLAKNKHTKEVLTYEKTKPEFYNVCVTDNYVRVFNEQSEQYIQKLPSK* 
 *HLKSSVKPLS*LFS*LGRVSGEDTLTETTKLLKLS*AMRS**VLLKMLIRKD*LH*APITKMI*KMCLLRISTQKRY*HMKKLSQSFIMFVLLIIMSGCSTNKVSSTFKNSQVS


(None, '\n', 'length:', 86252)

In [541]:
phage81 = test_spanin(data['phage_acc'][81], [54218,54575], test_cds = [54219,54575], strand='POSITIVE')
phage81

(None, '\n', 'length:', 23931)

In [542]:
data.iloc[81]

host                                                     Bdellovibrio
host_taxid                                                        662
phage_acc                                                 NC_015721.1
protein_acc                                                       NaN
protein_gi                                                   33882695
cds                                                               NaN
strand                                                            NaN
function                                                      OSPANIN
spanin_type                                                       ovl
property_feature                                          LIPO:25..28
sequence            MQKNIRMPRLGPIKFALLLILSIGIISCAGVPKFPKDIKVWETAYL...
ncbi_sequence                                                     NaN
same                                                              NaN
Name: 81, dtype: object

In [543]:
data['sequence'][81]

'MQKNIRMPRLGPIKFALLLILSIGIISCAGVPKFPKDIKVWETAYLPPPKDVPRGTPGQWFCGEYEPTTKNITRAKDIKFKPVKDHDISKCVGVFGFKDKDFPAVLDWMDRVQNYYEDRLERCEAKK'

In [544]:
phage118 = test_spanin(data['phage_acc'][118], [56741,57115], test_cds = [56740,57115], strand='POSITIVE')
phage118

 *R*FLLSYYH*LY*QVAVPCRWLQV*QAP*AENNLVYQWTPRWVIEWLMLARTITQQ*MPKITKVSSQ*LLINPRSNSTVPNR*PYRNLIVSWKCYYIPLV*LVGCYLILLRCIQKLNRGSPNL 
 VKIISTILLSLALLTGCSAMQMATGIASTLSGEQPSLSVDAQMGDRVANVGENDNSTINAEDNEGILTVTSNKSEKQFNGAQQITIQESNSFVEMLLYSIGLIGWLLPNPSQMYSEVKSWFSKS*


(None, '\n', 'length:', 66316)

In [545]:
data.iloc[118]

host                                                           vibrio
host_taxid                                                        662
phage_acc                                                 NC_023863.2
protein_acc                                                       NaN
protein_gi                                                  657201235
cds                                                               NaN
strand                                                            NaN
function                                                      USPANIN
spanin_type                                                       USP
property_feature                                          LIPO:14..17
sequence            MKIISTILLSLALLTGCSAMQMATGIASTLSGEQPSLSVDAQMGDR...
ncbi_sequence                                                     NaN
same                                                              NaN
Name: 118, dtype: object

In [546]:
data['sequence'][118]

'MKIISTILLSLALLTGCSAMQMATGIASTLSGEQPSLSVDAQMGDRVANVGENDNSTINAEDNEGILTVTSNKSEKQFNGAQQITIQESNSFVEMLLYSIGLIGWLLPNPSQMYSEVKSWFSKS'

In [547]:
# Get cds data in a column in the format of
# rohit:XX..YY
# make a notes column and copy my notes into it for questionable spanins

In [548]:
# for other category, need to check and verify the sequence is what the cds says it is (cat1 other csv file)
# it might be the negative strand (or some of them might be)
# check if sequence is good, if it is not, check the other opposite strand
# if neither of those are right, then mark it to be looked at further by curtis/jolene

* Phage 2 = called in NCBI now as hypothetical protein with CDS 134887:135240. Identical to rohit's sequence. Negative strand.
* Phage 5 = called in NCBI now as "Rz" with CDS 17544:18005. Identical to rohit's sequence. Negative strand.
* Phage 25 = called in NCBI now as hypothetical protein with CDS 41130:41510. Identical to rohit's sequence. However 41129:41510 is what properly translates in biopython. Pos strand.
* Phage 40 = called in NCBI now as hypothetical protein with CDS 5432:5779. Identical to rohit's sequence. However 5431:5779 is what properly translates in biopython. Pos strand.
* Phage 41 = called in NCBI now as 18.7 with CDS 35510:35761. Identical to rohit's sequence. However 35509:35761 is what properly translates in biopython. Pos strand.
* Phage 42 = called in NCBI as conserved hypothetical protein with CDS 134591:134890. Identical to rohit's sequence. Negative strand.
* Phage 74 = called in NCBI as hypothetical protein with CDS 54219:54575. Identical to rohit's sequence. 54218:54575 is what properly translates. Pos strand.
* Phage 81 = Rohit's call not found anywhere in the NCBI proteins. No spanins or hypothetical lipoproteins called in the NCBI data either. 
* Phage 114 = called in NCBI as hypothetical protein with CDS 5673:5942. Identical to rohit's sequence but 5672:5942 is what properly translates. Pos strand.
* Phage 118 = called in NCBI as hypothetical protein with CDS 56741:57115. Identical to rohit's sequence but 56740:57115 is what properly trnaslates. Pos strand.

In [549]:
data3 = data.copy()
data3['new_cds'] = data2['new_cds']
data3.head(3)

host  host_taxid    phage_acc     protein_acc  protein_gi  \
0       Aeromonas         642  NC_020879.1  YP_007677913.1   472438133   
1       Aeromonas         642  NC_019543.1  YP_007010874.1   423262275   
2  Enterobacteria         547     AP011113             NaN    26042140   

              cds    strand function spanin_type property_feature  \
0  126555..126789  NEGATIVE  ISPANIN         sep        TMD:7..29   
1  124571..124826  NEGATIVE  ISPANIN         sep        TMD:7..29   
2             NaN       NaN  ISPANIN         sep        TMD:5..24   

                                            sequence  \
0  MLLTLSSLLSWLKSNALCIIIMVLMAIMMKNQHDEISTLKTSLESM...   
1  MLLTLSSLLSWLKSNALYIIIMVLMAIMMKNQHDEISTLKNSLESM...   
2  MHVSNFTAGLLLLVIAFGGTSIILKNKVERLETSVVEITKTANENA...   

                                       ncbi_sequence same           new_cds  
0  MKNQHDEISTLKTSLESMKSFQTKSYENAKPVTEALLKSPKATKQM...    N  [126555, 126873]  
1  MVLMAIMMKNQHDEISTLKNSLESMKSFQTKSYENAKPVTEALLKS...    N  [124571, 124889]  
2                                                NaN  NaN               NaN

In [550]:
data3['new_cds'][2] = [134887, 135240]
data3['new_cds'][5] = [17544, 18005]
data3['new_cds'][25] = [41130, 41510]
data3['new_cds'][40] = [5432, 5779]
data3['new_cds'][41] = [35510, 35761]
data3['new_cds'][42] = [134591, 134890]
data3['new_cds'][74] = [54219, 54575]
data3['new_cds'][114] = [5673, 5942]
data3['new_cds'][118] = [56741, 57115]

C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so w

In [551]:
data3['new_cds'][45] = [99401,99680]
data3['new_cds'][79]= [26592,26849]
data3['new_cds'][99] = [4835, 5066]

C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so w

* Phage 3   = Phage 3 is a mess. Genome was updated to NC_024142.2. New cds in ncbi is 12842..13066, however their CDS in ncbi is not a multiple of three. AND I get a different sequence when I translate that cds (probably because the genome in NCBI is a slightly different size than the one efetch gets me...?)
* Phage 4 = Deletion of 'I' in rohit's sequence vs NCBI's. 89203:89547 is the ncbi cds
* Phage 23 = Rohit's annotation appears to be incorrect here (CDS is 167035,167455 however). He ends one amino acid before the stop codon, while NCBI's annotation correctly goes to the stop codon
* Phage 34 = 137084:137234, NC_024794.1 (manually)
* Phage 46 =  35459:35684, NC_023576.1
* Phage 50 =  36184:36415
* Phage 69 = This one is a mess too. The protein_acc and protein_gi are out of date and do not match anything in ncbi. They appear to have re-annotated it  by shifting the cds from 36645:36921 to 36550:37000 - which is completely different from rohit's annotation and their original annotation - not even in the same reading frame
* Phage 72 = 38503:38710 **but** I find rohit's annotation a little questionable here. He's missing a huge part of the orf
* Phage 98 = 38776:39058
* Phage 115 = My guess is that they annotated different genes here because the sequence is completely different

In [571]:
data3['new_cds'][4] = [89203, 89547]
data3['new_cds'][23] = [167035, 167455]
data3['new_cds'][34] = [137084, 137234]
data3['new_cds'][46] = [35459, 35684]
data3['new_cds'][50] = [36184, 36415]
data3['new_cds'][72] = [38503, 38710]
data3['new_cds'][81] = [0,0]
data3['new_cds'][98] = [38776, 39058]

C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\jcham\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so w

In [553]:
# Convert cds data in a column to the format of
# rohit:XX..YY

In [573]:
data3['new_cds'].apply(lambda x: str(x[0]) + '..' + str(x[1]))

0      126555..126873
1      124571..124889
2      134887..135240
3        12559..13066
4        89203..89547
            ...      
117      44709..45078
118      56741..57115
119      20740..21127
120         629..1010
121      41449..41839
Name: new_cds, Length: 122, dtype: object

In [574]:
# data3.to_csv('category_I_sequence_diff_JSC.csv')

In [ ]:
os.getcwd()

In [572]:
data3[data3['new_cds'].isnull() == True]

Empty DataFrame
Columns: [host, host_taxid, phage_acc, protein_acc, protein_gi, cds, strand, function, spanin_type, property_feature, sequence, ncbi_sequence, same, new_cds]
Index: []